src/TreeAutomata.hs

```diff
@@ -94,70 +112,194 @@ instance Hashable Rhs where
+    f n = do r <- get
+             unless (Set.member n r) $
+               when (any (all (`Set.member` r) . rhsNames) (case Map.lookup n p of Just x -> x)) $ do
+                 put (Set.insert n r)
+                 sequence_ [f x | x <- Map.findWithDefault [] n invMap]
+  return (Grammar s (Map.map filterProds (Map.filterWithKey (\k _ -> Set.member k nonEmpty) p)))
+
+-- | Removes productions that are not reachable form the start
+dropUnreachable :: GrammarBuilder -> GrammarBuilder
+dropUnreachable g = do
+  Grammar s p <- g
+  let
+    reachables = execState (f s) Set.empty
+    f :: Name -> State (Set.Set Name) ()
+    f n = do r <- get
+             unless (Set.member n r) $ do
+               put (Set.insert n r)
+               sequence_ [mapM_ f (rhsNames x) | x <- fromMaybe (error ("Name " ++ show n ++ " not in the grammar")) (Map.lookup n p)]
+  return $ Grammar s (Map.filterWithKey (\k _ -> Set.member k reachables) p) where
+
+-- | Removes useless productions.
+-- We drop unreachable first because that plays better with laziness.
+-- But we also drop unreachable after droping empty, because the empty may lead to unreachable.
+normalize :: GrammarBuilder -> GrammarBuilder
+normalize = dropUnreachable . dropEmpty . dropUnreachable
+
+-- | Removes all nonproductive non-terminals from the given grammar.
+removeUnproductive :: GrammarBuilder -> GrammarBuilder
+removeUnproductive g = do
+  Grammar start prods <- g
+  prodNs <- productive g
+  return (Grammar start (Map.filterWithKey (\k _ -> k `Set.member` prodNs) prods))
+
+-- | Returns all productive nonterminals in the given grammar.
+productive :: GrammarBuilder -> State Int (Set.Set Name)
+productive g = do
+  Grammar _ prods <- g
+  let
+    p = Set.fromList [ n | (n, rhss) <- Map.toList prods, producesConstant rhss]
+    producesConstant :: [Rhs] -> Bool
+    producesConstant = isJust . find (\r -> case r of Ctor _ [] -> True; _ -> False)
+    filter :: [Rhs] -> Set.Set Name -> Bool
+    filter rhss p = case rhss of
+      (Ctor _ args : rhss) -> if and (map (`Set.member` p) args) then True else filter rhss p
+      (Eps nonterm : rhss) -> if Set.member nonterm p then True else filter rhss p
+      [] -> False
+    go :: Map.Map Name [Rhs] -> State (Set.Set Name) ()
+    go prods = do p <- get
+                  let p' = Set.union p $ Set.fromList [ n | (n, rhss) <- Map.toList prods, filter rhss p ]
+                  put p'
+                  if p == p' then return () else go prods
+  return (execState (go prods) p)
+
+-- | Returns a list of grammars, where each grammar has a non-terminal
+-- symbol from the given grammar as start symbol.
+permutate :: GrammarBuilder -> State Int [GrammarBuilder]
+permutate g = do
+   Grammar _ ps <- g
+   return (map (\n -> return (Grammar n ps)) (Map.keys ps))
+
+-- | Returns true iff the grammar can construct the given constant.
+produces :: GrammarBuilder -> Name -> State Int Bool
+produces g n = do
+  Grammar s prods <- g
```

Comment:
You are generating unique names of `g` twice. `productive` should operate on the grammar that `g` produced.